### ISSUE TITLE
Reloading with USR1 signal when server is not "ready" prevents the server to be able to become ready

### ISSUE THREAD

**Issue by vitalyvb** [miekg/coredns#5481]
**What happened**:

Single-node Kubernetes cluster boots, kube-apiserver and CoreDNS are starting (not exactly at the same time). Early at start CoreDNS config is updated and CoreDNS is reloaded via "pkill -e -USR1 coredns".

CoreDNS gets stuck in 'Still waiting on: "kubernetes"' state and DNS server never gets ready (readiness check always fails).

**What you expected to happen**:

After receiving USR1 signal, configuration should be reloaded and DNS server should be able to become ready.

**How to reproduce it (as minimally and precisely as possible)**:

1. CoreDNS should be waiting for "kubernetes" plugin to be ready (or perhaps any other plugin)
2. Send USR1 signal

**Anything else we need to know?**:

CoreDNS memory usage increases when it handles thousands of USR1 signals.

It looks like "plugins" list of "ready" plugin never gets emptied, but plugins are added when configuration is reloaded:
https://github.com/coredns/coredns/blob/390e88637e4307e37b494a70d58e9a2871c5f6bf/plugin/ready/setup.go#L27-L45
https://github.com/coredns/coredns/blob/390e88637e4307e37b494a70d58e9a2871c5f6bf/plugin/ready/ready.go#L69-L81

**Environment**:
Kubernetes 1.22.6

- the version of CoreDNS:
1.8.7
1.8.4
1.6.2

- Corefile:
```
    .:53 {
        log . {
          class denial error
        }
        errors
        health
        ready
        kubernetes cluster.local in-addr.arpa ip6.arpa {
          pods insecure
          fallthrough in-addr.arpa ip6.arpa
        }
        prometheus :9153
        forward . /host-cfg/resolv.conf {
          prefer_udp
          max_fails 2
          expire 3s
          policy sequential
        }
        cache 30
        loop
        reload
        loadbalance
    }
```

- logs, if applicable:
```
[INFO] plugin/ready: Still waiting on: "kubernetes"
[INFO] plugin/ready: Still waiting on: "kubernetes"
[INFO] plugin/ready: Still waiting on: "kubernetes"
[INFO] SIGUSR1: Reloading
[INFO] Reloading
[WARNING] plugin/kubernetes: starting server with unsynced Kubernetes API
.:53
CoreDNS-1.8.4
linux/amd64, go1.16.4, 053c4d5
[INFO] plugin/ready: Still waiting on: "kubernetes,kubernetes"
[WARNING] plugin/kubernetes: starting server with unsynced Kubernetes API
[INFO] Reloading complete
[INFO] plugin/ready: Still waiting on: "kubernetes,kubernetes"
[INFO] plugin/ready: Still waiting on: "kubernetes,kubernetes"
[INFO] plugin/ready: Still waiting on: "kubernetes"
[INFO] plugin/ready: Still waiting on: "kubernetes"
[INFO] plugin/ready: Still waiting on: "kubernetes"
...
```
- OS (e.g: `cat /etc/os-release`):
coredns/coredns image

- Others:
n/a

**Comment by chrisohaver** [comment]
> ... CoreDNS gets stuck in 'Still waiting on: "kubernetes"' state and DNS server never gets ready (readiness check always fails).

This happens when the Kubernetes client cannot connect to the k8s API for some reason. Example reasons for this could be that the pod has no route to the API, the API isn't ready, or the pod doesn't have the proper auth/certs. 

> **What you expected to happen:**
> After receiving USR1 signal, configuration should be reloaded and DNS server should be able to become ready.

The ready plugin by design will not return ready until the the Kubernetes successfully connects and syncs data with the k8s API.  So, if there is something preventing that from happening , the ready plugin will not report ready.

There were a few releases of CoreDNS where API connection errors were not logged at startup.
This was fixed in 1.9.1.  If you update that version or later, you may get a better idea of why the kubernetes plugin is unable to sync with the API.

> It looks like "plugins" list of "ready" plugin never gets emptied, but plugins are added when configuration is reloaded:

Thanks for catching that. It should be fixed. Although I don't think it relates to the issue you see (CoreDNS unready).

**Comment by vitalyvb** [comment]
With 1.9.3:
```
[INFO] plugin/kubernetes: waiting for Kubernetes API before starting server
[INFO] plugin/kubernetes: waiting for Kubernetes API before starting server
[INFO] plugin/kubernetes: waiting for Kubernetes API before starting server
[INFO] plugin/kubernetes: waiting for Kubernetes API before starting server
[INFO] plugin/kubernetes: waiting for Kubernetes API before starting server
[WARNING] plugin/kubernetes: starting server with unsynced Kubernetes API
.:53
CoreDNS-1.9.3
linux/amd64, go1.18.2, 45b0a11
[INFO] plugin/kubernetes: waiting for Kubernetes API before starting server
[INFO] plugin/kubernetes: waiting for Kubernetes API before starting server
[INFO] plugin/kubernetes: waiting for Kubernetes API before starting server
[INFO] plugin/kubernetes: waiting for Kubernetes API before starting server
[INFO] plugin/ready: Still waiting on: "kubernetes,kubernetes"
[WARNING] plugin/kubernetes: starting server with unsynced Kubernetes API
[INFO] 127.0.0.1:40906 - 33869 "HINFO IN 8421009060179317439.3566712214548512059. udp 57 false 512" NXDOMAIN qr,rd,ra 132 0.024966979s
[INFO] 127.0.0.1:49085 - 23370 "HINFO IN 8421009060179317439.3566712214548512059. udp 57 false 512" - - 0 6.00264073s
[ERROR] plugin/errors: 2 8421009060179317439.3566712214548512059. HINFO: read udp 198.51.100.3:49594->192.0.2.249:53: i/o timeout
[INFO] 127.0.0.1:38804 - 46612 "HINFO IN 4262351901766538315.8283857908586835412. udp 57 false 512" - - 0 6.002228976s
[ERROR] plugin/errors: 2 4262351901766538315.8283857908586835412. HINFO: read udp 198.51.100.3:45565->192.0.2.249:53: i/o timeout
[INFO] Reloading complete
[INFO] 127.0.0.1:38388 - 36275 "HINFO IN 4262351901766538315.8283857908586835412. udp 57 false 512" NXDOMAIN qr,rd,ra 132 0.023915179s
[INFO] plugin/ready: Still waiting on: "kubernetes,kubernetes"
[INFO] plugin/ready: Still waiting on: "kubernetes,kubernetes"
[INFO] plugin/ready: Still waiting on: "kubernetes"
[INFO] plugin/ready: Still waiting on: "kubernetes"
[INFO] plugin/ready: Still waiting on: "kubernetes"
...
```

It "recovers" after "killall coredns".

On the other hand, a rare successful reload when server boots looks like this:
```
[INFO] plugin/ready: Still waiting on: "kubernetes"
[INFO] plugin/ready: Still waiting on: "kubernetes"
[INFO] plugin/ready: Still waiting on: "kubernetes"
[WARNING] plugin/kubernetes: starting server with unsynced Kubernetes API
.:53
[INFO] plugin/reload: Running configuration MD5 = fa688ba404fdf2c46a7f4f9d3007ed7d
CoreDNS-1.8.4
linux/amd64, go1.16.4, 053c4d5
[INFO] plugin/ready: Still waiting on: "kubernetes"
[INFO] SIGUSR1: Reloading
[INFO] Reloading
[INFO] 127.0.0.1:39884 - 62708 "HINFO IN 5477640084865121922.3994135939565831056. udp 57 false 512" NOERROR - 0 6.000823948s
[ERROR] plugin/errors: 2 5477640084865121922.3994135939565831056. HINFO: read udp 198.51.100.4:37454->192.0.2.249:53: i/o timeout
[INFO] 127.0.0.1:39074 - 53702 "HINFO IN 9175120288254922384.116661857449248682. udp 56 false 512" NOERROR - 0 2.001013863s
[ERROR] plugin/errors: 2 9175120288254922384.116661857449248682. HINFO: read udp 198.51.100.4:43583->192.0.2.249:53: i/o timeout
[INFO] 127.0.0.1:42205 - 58859 "HINFO IN 5477640084865121922.3994135939565831056. udp 57 false 512" NOERROR - 0 6.001536182s
[ERROR] plugin/errors: 2 5477640084865121922.3994135939565831056. HINFO: read udp 198.51.100.4:35554->192.0.2.249:53: i/o timeout
[INFO] 127.0.0.1:52194 - 16365 "HINFO IN 5477640084865121922.3994135939565831056. udp 57 false 512" NXDOMAIN qr,rd,ra 132 0.032992949s
[INFO] 127.0.0.1:36172 - 3243 "HINFO IN 9175120288254922384.116661857449248682. udp 56 false 512" NXDOMAIN qr,rd,ra 131 0.027991566s
[INFO] 127.0.0.1:46429 - 58302 "HINFO IN 5477640084865121922.3994135939565831056. udp 57 false 512" NOERROR - 0 4.001369105s
[ERROR] plugin/errors: 2 5477640084865121922.3994135939565831056. HINFO: read udp 198.51.100.4:51804->192.0.2.249:53: i/o timeout
[INFO] Reloading complete
... dns works ...
```

I really don't know if the problem is in "ready" plugin or some other list, but from the log it looks like that USR1 is received while it's waiting "kubernetes" and after the reload it waits for two kuberneteses: "kubernetes,kubernetes". It also looks like one of those kuberneteses becomes ready and it waits for the other "kubernetes", but config file always has only one instance of kubernetes.

**Comment by chrisohaver** [comment]
> It also looks like one of those kuberneteses becomes ready and it waits for the other "kubernetes", but config file always has only one instance of kubernetes.

Ah, yes - I think it can be the cause - the old instance may still be unable to connect for whatever reason (perhaps the k8s client reads certain creds at startup only).

Fix here: #5492

**Comment by rtheis** [comment]
We are hitting a similar problem where some CoreDNS pods are running and ready but contain stale or no data for Kubernetes services.  This can then lead to DNS troubles for our in cluster apps.  A restart of the problem CoreDNS pods fixes the problem.  Problem CoreDNS pods normally contain logs similar to the following:

```
[INFO] plugin/ready: Still waiting on: "kubernetes"

...

[INFO] plugin/reload: Running configuration MD5 = <MD5>

...
```

I found this issue and it certainly seems related to what we are seeing.  If you disagree, we can open a new issue.  Thanks.
